test/tests/functional/pbs_hook_endjob.py

```diff
@@ -0,0 +1,1238 @@
+            # TERMINATED (91).  otherwise, verify that the substate is set to
+            # FINISHED (92).
+            if jid in self.delete_failed_job_ids:
+                substate = 93
+            elif jid in self.deleted_job_ids:
+                substate = 91
+            else:
+                substate = 92
+            self.server.expect(
+                JOB, {'job_state': 'F', 'substate': substate}, extend='x',
+                id=jid)
+            # if the job was deleted without the force flag and the moms were
+            # stopped and not restarted, then an accounting 'E' record will not
+            # be immediately written.
+            if not (jid in self.delete_failed_job_ids and self.moms_stopped):
+                self.server.accounting_match(
+                    "E;%s;" % (jid,),
+                    starttime=int(self.log_start_time),
+                    n='ALL', max_attempts=1)
+            self.ended_job_ids.add(jid)
+
+    def resv_submit(
+            self,
+            user=TEST_USER,
+            nchunks=resv_default_nchunks,
+            ncpus=resv_default_ncpus,
+            resv_start_time=None,
+            resv_end_time=None,
+            resv_attrs={}):
+        start_time = resv_start_time or int(time.time()) + \
+            self.resv_start_delay
+        end_time = resv_end_time or start_time + self.resv_duration
+        a = {}
+        a['Resource_List.select'] = str(nchunks) + ':ncpus=' + str(ncpus)
+        a['Resource_List.place'] = 'free'
+        a['reserve_start'] = start_time
+        a['reserve_end'] = end_time
+        a.update(resv_attrs)
+        resv = Reservation(user, a)
+        self.resv_id = self.server.submit(resv)
+        self.resv_start_time = start_time
+
+    def resv_verify_confirmed(self):
+        a = {'reserve_state': (MATCH_RE, 'RESV_CONFIRMED|2')}
+        self.server.expect(RESV, a, id=self.resv_id)
+        self.resv_queue = self.resv_id.split('.')[0]
+        self.server.status(RESV, 'resv_nodes')
+
+    def resv_verify_started(self):
+        self.logger.info('Sleeping until reservation starts')
+        self.server.expect(
+            RESV, {'reserve_state': (MATCH_RE, 'RESV_RUNNING|5')},
+            id=self.resv_id,
+            offset=self.resv_start_time - int(time.time() + 1))
+
+    def moms_start(self, *args, **kwargs):
+        for mom in self.moms.values():
+            mom.start(*args, **kwargs)
+        self.moms_stopped = False
+        if not self.scheduling_enabled:
+            self.server.manager(MGR_CMD_SET, SERVER, {'scheduling': 'True'})
+            self.scheduling_enabled = True
+
+    def moms_stop(self, *args, **kwargs):
+        if self.scheduling_enabled:
+            self.server.manager(MGR_CMD_SET, SERVER, {'scheduling': 'False'})
+            self.scheduling_enabled = False
+        for mom in self.moms.values():
+            mom.stop(*args, **kwargs)
+        self.moms_stopped = True
+
```

Comment:
The `TestHookEndJob` class level attributes are all constants that are largely used as default argument values.  All of the attributes modified by the helper methods in the `TestHookEndJob` class are part of the object instantiated by nose not the class itself.  These attributes are used to track state which can change within each helper method as the test progresses, so passing them as arguments doesn't seem like an option.  The attributes are initialized, and thus reset, at the beginning of the `run_test_func()` method, which is used by all of the test methods in the class.  If you think it would improve code readability, we can move the initialization of those attributes into a `setUp()` method and add a `tearDown()` method to clear them.

Reply:
setUp() is called before each job.  If you reinitialize them in setUp() you should be fine.  You don't need setUp() and tearDown() both.